File: web/lib/artifact-store.ts
// Web-side ArtifactStore (docs/branch-c.md). The agent's Python store writes
// the row at plan-lock; this class reads it and owns photos/keepsake state.
// Never imported by the agent — the table IS the seam.
import { db } from "./db";

export interface Place {
  name: string;
  area?: string | null;
  url?: string | null;
  note?: string | null;
}

export interface Track {
  title: string;
  artist: string;
  url?: string | null;
}

export interface HangoutArtifact {
  plan_id: string;
  place: Place;
  time: string;
  attendees: string[];
  playlist: Track[];
  photos: string[];
  created_at: string;
  isKeepsake: boolean;
}

interface Row {
  plan_id: string;
  place: string;
  time: string;
  attendees: string;
  playlist: string;
  photos: string;
  created_at: string;
}

function parse(row: Row): HangoutArtifact {
  const photos = JSON.parse(row.photos) as string[];
  return {
    plan_id: row.plan_id,
    place: JSON.parse(row.place),
    time: row.time,
    attendees: JSON.parse(row.attendees),
    playlist: JSON.parse(row.playlist),
    photos,
    created_at: row.created_at,
    isKeepsake: photos.length > 0,
  };
}

export class ArtifactStore {
  create(
    plan: { plan_id: string; place: Place; time: string; attendees: string[] },
    playlist: Track[]
  ): HangoutArtifact {
    db()
      .prepare(
        "INSERT INTO artifacts (plan_id, place, time, attendees, playlist) VALUES (?, ?, ?, ?, ?)"
      )
      .run(
        plan.plan_id,
        JSON.stringify(plan.place),
        plan.time,
        JSON.stringify(plan.attendees),
        JSON.stringify(playlist)
      );
    return this.get(plan.plan_id)!;
  }

  get(planId: string): HangoutArtifact | null {
    const row = db().prepare("SELECT * FROM artifacts WHERE plan_id = ?").get(planId) as
      | Row
      | undefined;
    return row ? parse(row) : null;
  }

  list(): HangoutArtifact[] {
    const rows = db()
      .prepare("SELECT * FROM artifacts ORDER BY created_at DESC, rowid DESC")
      .all() as Row[];
    return rows.map(parse);
  }

  addPhotos(planId: string, urls: string[]): void {
    const existing = this.get(planId);
    if (!existing) throw new Error(`no artifact ${planId}`);
    db()
      .prepare("UPDATE artifacts SET photos = ? WHERE plan_id = ?")
      .run(JSON.stringify([...existing.photos, ...urls]), planId);
  }
}

File: src/agent/artifact_store.py
"""T9: agent-side ArtifactStore — writes the artifacts row at plan-lock (FR17).

C's web app reads/writes the same table; this class is never imported by web/.
"""

import json
import sqlite3

from src.contracts import Candidate, FinalPlan, HangoutArtifact, Track


class SqliteArtifactStore:
    def __init__(self, db_path: str):
        self._db_path = db_path

    def _conn(self) -> sqlite3.Connection:
        return sqlite3.connect(self._db_path)

    async def create(self, plan: FinalPlan, playlist: list[Track]) -> HangoutArtifact:
        with self._conn() as conn:
            conn.execute(
                "INSERT INTO artifacts (plan_id, place, time, attendees, playlist)"
                " VALUES (?, ?, ?, ?, ?)",
                (
                    plan.plan_id,
                    plan.place.model_dump_json(),
                    plan.time.isoformat(),
                    json.dumps(plan.attendees),
                    json.dumps([t.model_dump() for t in playlist]),
                ),
            )
        return HangoutArtifact(
            plan_id=plan.plan_id,
            place=plan.place,
            time=plan.time,
            attendees=plan.attendees,
            playlist=playlist,
        )

    async def get(self, plan_id: str) -> HangoutArtifact | None:
        row = self._conn().execute(
            "SELECT place, time, attendees, playlist, photos, created_at"
            " FROM artifacts WHERE plan_id = ?",
            (plan_id,),
        ).fetchone()
        if row is None:
            return None
        return HangoutArtifact(
            plan_id=plan_id,
            place=Candidate.model_validate_json(row[0]),
            time=row[1],
            attendees=json.loads(row[2]),
            playlist=[Track(**t) for t in json.loads(row[3])],
            photos=json.loads(row[4]),
            created_at=row[5],
        )

    async def add_photos(self, plan_id: str, urls: list[str]) -> None:
        with self._conn() as conn:
            (existing,) = conn.execute(
                "SELECT photos FROM artifacts WHERE plan_id = ?", (plan_id,)
            ).fetchone()
            conn.execute(
                "UPDATE artifacts SET photos = ? WHERE plan_id = ?",
                (json.dumps(json.loads(existing) + urls), plan_id),
            )
